### ISSUE TITLE
GCC Examples Needed

### ISSUE THREAD

**Issue by oamike** [3MFConsortium/lib3mf-examples#3]
Attempting to build any of the GCC examples fails in MacOS using gcc.
Has anyone been successful at building on Mac?

**Comment by ofloveandhate** [comment]
same error here.  i think we are intended to modify the Makefile, but without any documentation whatsoever, this is not something i really want to engage in.

**Comment by ofloveandhate** [comment]
modifying the Makefile, line 13 
from
`INC =`
to
`INC = -I/usr/local/include/lib3mf`
fixes the include problem, but yields this one:

```
g++ -Wall -fexceptions -g -D__GCC -I/usr/local/include/lib3mf -I../../include -c Converter.cpp -o obj/Debug/Converter.o
Converter.cpp:151:12: error: no matching function for call to 'lib3mf_createmodel'
        hResult = lib3mf_createmodel(&pModel);
                  ^~~~~~~~~~~~~~~~~~
../../include/NMR_DLLInterfaces.h:105:32: note: candidate function not viable: requires 2 arguments, but 1 was provided
                LIB3MF_DECLSPEC LIB3MFRESULT lib3mf_createmodel(_Outptr_ PLib3MFModel ** ppModel, _In_ BOOL bInitialize);
                                             ^
1 error generated.
make: *** [obj/Debug/Converter.o] Error 1
```

**Comment by t-paul** [comment]
This should be `lib3mf_createmodel(&pModel, true);` now, I think.

**Comment by ofloveandhate** [comment]
yep, can confirm.  then fails at link time, because the library extension is hardcoded as `.so`.  about to test `.dylib` on mac

**Comment by ofloveandhate** [comment]
changing lines 18 and 19 in Makefile get to link, but still fails.  new lines:
```
LIB = -l3mf
LDFLAGS = -L/usr/local/lib
```

but linker error for incorrect architecture:

```
g++  -o bin/Debug/Converter obj/Debug/Converter.o  -L/usr/local/lib -l3mf 
ld: warning: ignoring file /usr/local/lib/lib3mf.dylib, file was built for i386 which is not the architecture being linked (x86_64): /usr/local/lib/lib3mf.dylib
Undefined symbols for architecture x86_64:
  "_lib3mf_createmodel", referenced from:
      _main in Converter.o
  "_lib3mf_getinterfaceversion", referenced from:
      _main in Converter.o
  "_lib3mf_getlasterror", referenced from:
      _main in Converter.o
  "_lib3mf_model_queryreader", referenced from:
      _main in Converter.o
  "_lib3mf_model_querywriter", referenced from:
      _main in Converter.o
  "_lib3mf_reader_readfromfileutf8", referenced from:
      _main in Converter.o
  "_lib3mf_release", referenced from:
      _main in Converter.o
  "_lib3mf_writer_writetofileutf8", referenced from:
      _main in Converter.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [out_debug] Error 1

```

so, i have to re-compile lib3mf, changing the architecture.  this mucking around in Makefiles is not so pleasant.  Have you considered CMake, which works on Windows for generating VS project files, as well as the *nix platforms?  This would really help clean this build/ install mess up.

**Comment by ofloveandhate** [comment]
using Makefile.Darwin in lib3mf doesn't work:
```
make -f Makefile.Darwin  install
make: *** No rule to make target `install'.  Stop.
```
similarly if i put target before the `-f` flag.

**Comment by t-paul** [comment]
It's not meant to be called directly. The system specific files are included from the main `Makefile`.

**Comment by ofloveandhate** [comment]
i respectfully ask for build instructions.  are they in that pdf at root of lib3mf?  i didn't see them.

**Comment by ofloveandhate** [comment]
ok, i see how the correct Makefile.system is included into Makefile.  But I am still getting incorrect architecture from the resulting .dylib.

**Comment by t-paul** [comment]
Looks like the Darwin file compiles a 32-bit version by default (when just using `make`).

Running `make CC="gcc -m64" CXX="g++ -m64" LD=g++ -m64"` might work.

**Comment by ofloveandhate** [comment]
ok, it made with replacing those.  i did it in the Makefile.  

I feel like those kinds of details are why a configure script, or some other build utility, is really important.  i just dislike messing with Makefiles, or having to mess with `CC` etc.  there are no help messages available from the command line, no feedback other than error messages can be hard to decode, especially for the inexperienced, who are the easiest to lose as users.  and these are all issues that the build system can take care of for us.  

i got a few warnings from the compiler.  not fully posted here; a separate issue would be more appropriate.

* `../../Source/Common/NMR_StringUtils.cpp:85:16: warning: comparison of constant 9223372036854775807 with expression of type 'nfInt32' (aka 'int') is
      always false [-Wtautological-constant-out-of-range-compare]
                if ((nResult == LONG_MAX) || (nResult == LONG_MIN))`
* `../../Source/Common/NMR_StringUtils.cpp:85:41: warning: comparison of constant -9223372036854775808 with expression of type 'nfInt32' (aka 'int') is
      always false [-Wtautological-constant-out-of-range-compare]
                if ((nResult == LONG_MAX) || (nResult == LONG_MIN))`
* and similarly at lines 108, 248, 271.

i got a warning from ld:
```
g++  -o bin/Release/Converter obj/Release/Converter.o  -L/usr/local/lib -s -l3mf 
ld: warning: option -s is obsolete and being ignored
```

the usage message for Converter is Windows-specific:

```
Usage: 
Convert 3MF to STL: Converter.exe model.3mf
Convert STL to 3MF: Converter.exe model.stl
```

have not fed it a model yet. will post back once i confirm that it at least doesn't die during conversion for a simple model.  if i encounter further problems i will raise a new issue, since this issue is kinda solved.  finally, once i confirm converter works, we should summarize what i did to get this compiled on my mac.

**Comment by ofloveandhate** [comment]
`Converter` ran correctly, on a model generated from numerical data, fed through Blender's solidify modifier, and then through the netfabb.azurewebsites.net, which has migrated to
https://tools3d.azurewebsites.net/ and now produces 3mf files. this was my motivation for engaging in this venture.  went from 249kb to 1.2 mb.  input .3mf file attached, in .zip because github doesn't support that file type
[dingdong-dingdong_smooth_repaired.3mf.zip](https://github.com/3MFConsortium/lib3mf-examples/files/501452/dingdong-dingdong_smooth_repaired.3mf.zip)
.



That is, it works.  thanks. summary in next post.

**Comment by ofloveandhate** [comment]
to get to Converted example to compile using XCode 8-provided Clang on OSX 10.11, i did the following:

1. adjust stuffs for lib3mf.  `make CC="gcc -m64" CXX="g++ -m64" LD="g++ -m64"`
2. `make install`
3. modify the Makefile, line 13 from `INC =` to `INC = -I/usr/local/include/lib3mf`.  also, `LIB = -l3mf`
`LDFLAGS = -L/usr/local/lib`, or pass these as variables to `make`.
4. Converter.cpp, add `true` to function argument at line 151.
5. `make`.  optionally, use `INC` `LIB` and `LDFLAGS` arguments, and don't modify Makefile.

there's no need to actually modify the Makefiles, you can just use arguments.  the defaults are wrong for this OS at this time.  pick your poison.

**Comment by edlutz** [comment]
The make files on the master branch assume a case insensitive file system, since they refer to the Include directory as "include" (lower case i). It looks like the GCC examples were not tested on unix-like environments like Linux.
